### PR TITLE
Transpose slice fix

### PR DESCRIPTION
**Pull request by shivadbhavsar**
This PR resolves the bug addressed in #1496. The issue points to several matchers in the simplify_algebra pass, however, upon further investigation, the simplify_algebra pass results in the correct program output. The issue was determined to be in the simplify_reshapes pass. 

List of updates:
1. find_split_reshape currently causes a segmentation fault if the reshaped split is the output of the program.
2. The transpose_slice simplification had a bug in its implementation causing the output of the compiled instructions to be a jumbled version of the true output.
3. Updated the existing simplify_reshape test to expect the correct modified program after the pass
4. Added a verify test for the same program found in simplify_reshapes_test.cpp (existing implementation caused this program to fail verify)

These changes allow us to run the full stable diffusion pipeline in fp32 mode correctly.

### CODE REVIEW

File: src/simplify_algebra.cpp [hunk 1]
```diff
@@ -1066,10 +1066,16 @@ struct find_split_reshape
         }
 
         std::vector<instruction_ref> vec_rsp(split_outputs.size());
-        std::transform(split_outputs.begin(), split_outputs.end(), vec_rsp.begin(), [](auto i) {
+        std::list<instruction> undef_inss{instruction(make_op("undefined"), {}, {})};
```
Comment (pfultz2) [inline src/simplify_algebra.cpp:1069]: This is the wrong way to insert undefined operator(as the instruction doesnt belong to any module in the program. Why are you inserting undefined operators in the first place? Why dont we just exit when output size is not 1?

File: test/simplify_reshapes_test.cpp [hunk 2]
```diff
@@ -1405,9 +1405,9 @@ TEST_CASE(transpose_slice_non_packed_axis)
     {
         auto x = m2.add_parameter("x", {migraphx::shape::float_type, {2, 384, 36, 64}});
         auto unsqueeze =
-            m2.add_instruction(migraphx::make_op("unsqueeze", {{"axes", {2}}, {"steps", {12}}}), x);
+            m2.add_instruction(migraphx::make_op("unsqueeze", {{"axes", {2}}, {"steps", {3}}}), x);
```
Comment (pfultz2) [inline test/simplify_reshapes_test.cpp:1408]: Why is this changed?

File: src/simplify_algebra.cpp (hunk 1 above)
Comment (shivadbhavsar) [reply src/simplify_algebra.cpp:1069]: The check for output size happens in the lambda function for each item in split_outputs and so adding the undefined instruction is to ensure that the following `same_ops` check fails causing it to exit

Comment (pfultz2) [reply src/simplify_algebra.cpp:1069]: I think it would be better to do `std::any_of` and check for the output size and then exit.

File: src/simplify_algebra.cpp [hunk 3]
```diff
@@ -1065,6 +1065,13 @@ struct find_split_reshape
             return;
         }
 
+        if(std::any_of(split_outputs.begin(), split_outputs.end(), [](auto i) {
+               return i->outputs().front()->outputs().size() != 1;
```
Comment (kahmed10) [inline src/simplify_algebra.cpp:1069]: As it's written right now, it's kind of confusing. The `std::transform` in L1076 helps explain better what these variables are supposed to represent (and also some of the asserts shouldn't be needed anymore with the addition you made). `i->outputs().front()` is a contiguous op. You may want to explicitly make a variable `cont` like in L1078. Also, we have an assert for `i->outputs().size() == 1` in L1077. I'm not sure if you want to make the function return where the size is != 1 and remove that assert as well.

Comment (shivadbhavsar) [reply src/simplify_algebra.cpp]: Updated to remove both asserts and also add a comment to make the intent behind the check more clear

**Review by kahmed10 — APPROVED** [review]

**Review by pfultz2 — APPROVED** [review]
